elastic/kibana
# [Telemetry] Settings Collector: redact sensitive reported values

The Usage Collector `stack_management` reports user changed settings.

This PR adds a new config to registered UI settings `sensitive: boolean`. The value of any UI setting marked as `sensitive` will be reported as a keyword `[REDACTED]` instead of the actual value. This hides the actual sensitive information while giving us some intelligence over which fields the users are interactive with the most.


## Changes
Core Features:
- A new `sensitive: true` config to the method `core.uiSettings.register`.

    ```
    uiSettings.register({
      [NEWS_FEED_URL_SETTING]: {
        name: i18n.translate('xpack.securitySolution.uiSettings.newsFeedUrl', {
          defaultMessage: 'News feed URL',
        }),
        value: NEWS_FEED_URL_SETTING_DEFAULT,

        description: i18n.translate('xpack.securitySolution.uiSettings.newsFeedUrlDescription', {
          defaultMessage: '<p>News feed content will be retrieved from this URL</p>',
        }),
        category: [APP_ID],
        requiresPageReload: true,
        sensitive: true,
        schema: schema.string(),
      },
    }),
    ```

- Add `isSensitive` method to `uiSettings` plugin contract.
- Values of UI Settings marked as sensitive will be reported reported as `[REDACTED]`

Schema updates:
- UI Settings collector schema is not up to date with all the ui settings we have and it is evergreen!
- Updated schema Typescript type `MakeSchemaFrom` to fail when the Usage TS type does not match the schema `type` property.
	  
	  ```
	  interface UsageStats {
	    someProp: number;
	  }
	  
	  const schema: MakeSchemaFrom<UsageStats> = {
	    someProp': { type: 'long' },
	  }
	  ```	  
	  
	Changing `long` to `keyword` for example will fail the type check
	(in IDE as well) and not just the `telemetry_check.ts`.



- Telemetry checker for UsageStats types now support the syntax `Array<T>` instead of only supporting `T[]`.

	  ```
	  interface UsageStats {
	    someProp: number[];
	    // this is now supported too!
	    someProp2: Array<number>;
	  }
	  ```

Tests:

- Added tests for the ui settings collector function.
- Add functional  tests to make sure all UI Settings are mapped in the schema.
- Add functional tests to make sure all sensitive UI settings are mapped as `keyword`s. This will help us make sure we dont accidentally report sensitive fields to telemetry since this will trigger a code review from the telemetry team.




## Current Sensitive UI Settings:

```
{
  'timelion:quandl.key'
  'securitySolution:defaultIndex'
  'securitySolution:newsFeedUrl'
  'xpackReporting:customPdfLogo'
  'notifications:banner'
  'timelion:graphite.url'
  'xpackDashboardMode:roles'
}
```

## Got pinged for a codereview?

It is most likely a change adding `sensitive: true` to one of the registered ui settings your team owns. Please filter by codeowners and check for the changes.

Closes https://github.com/elastic/kibana/issues/85811
